commands/channelmarshaler.go

```diff
@@ -34,11 +34,9 @@ func (cr *ChannelMarshaler) Read(p []byte) (int, error) {
 	}
 
 	n, err := cr.reader.Read(p)
-	if err != nil && err != io.EOF {
-		return n, err
-	}
-	if n == 0 {
+	if err == io.EOF {
 		cr.reader = nil
+		err = nil
```

Comment:
Are you sure we want to `nil` out the error in this case and not propagate the `io.EOF`? This is `Read` method after all.

Reply:
Had to think about that for a moment... yes. We're *concatenating* output so this is the *continue* case. The next time the user calls Read, we'll hit the `cr.reader == nil` case. If there are no more readers, we'll return the EOF.

Reply:
Ahh, right.

Reply:
Doesn't this create the case @kevina described in https://github.com/ipfs/go-ipfs/issues/4793#issuecomment-371372119 (where we return nothing and no error)?

Reply:
Good point. We should probably loop.